### PR TITLE
Fix destroying descriptors with UPDATE_AFTER_BIND

### LINKED ISSUE #3729: Wrong Validation Message when destroying ressource in descriptor set with partially bound flag
**Describe the Issue**
When using a descriptor set, created with the partially bound bit and destroying a ressource in that set wich is not 'dynamicly used' by a shader the validation layers give a false positive error message like below.
When turning on gpu assisted validation this false positive message dissapears indicating there is indeed a bug in the validation layers.

**Valid Usage ID**
creating a descriptor set with the VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT,
binding an image
using the set in a shader (but not the image, the image is in a descriptor array and is not used in the shader)
destroying the image that is not used.
cause this false validation message:
> VUID-vkDestroyBuffer-buffer-00922(ERROR / SPEC): msgNum: -464217071 - Validation Error: [ VUID-vkDestroyBuffer-buffer-00922 ] Object 0: handle = 0x4868e6000000005a, name = primitiveInfoBuffer, type = VK_OBJECT_TYPE_BUFFER; | MessageID = 0xe4549c11 | Cannot free VkBuffer 0x4868e6000000005a[primitiveInfoBuffer] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
>     Objects: 1
>         [0] 0x4868e6000000005a, type: 9, name: primitiveInfoBuffer
> [ERROR: Validation]
> Validation Error: [ VUID-vkDestroyBuffer-buffer-00922 ] Object 0: handle = 0x4868e6000000005a, name = primitiveInfoBuffer, type = VK_OBJECT_TYPE_BUFFER; | MessageID = 0xe4549c11 | Cannot free VkBuffer 0x4868e6000000005a[primitiveInfoBuffer] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)

**Environment:**
 - OS: Windows 10 
 - GPU: GTX 1080ti Driver: 511.23
 - SDK or header version if building from repo: 1.2.198.1, layers build from master: 58caea82d2cb199142f93664475fc413f03bd26e
 - Options enabled (synchronization, best practices, etc.):
 Core, Handle Wrapping, Object Lifetime, Stateless Parameter

**Additional context**
The set that is refered to above is very big four descriptor arrays with each 1024 descriptors. 
I am also using dynamic rendering wich is not fully supported yet.

### LINKED ISSUE #3835: DrawState validation raised on destroyed ACCELERATION_STRUCTURE_KHR despite UPDATE_AFTER_BIND
**Describe the Issue**
When an AccelerationStructure is bound to a descriptor set using `UPDATE_AFTER_BIND`, the validation layer reports an error when the binding is invalidated ~before~ between when it was recorded compared to when it was submitted. The update after bind flag should make this usage valid according to the spec, and the validation error is not raised in other areas of my app when the descriptor_type is `COMBINED_IMAGE_SAMPLER`.

The timeline of calls that lead to the error:

1. `vkUpdateDescriptorSets(acceleration_set[0] = acceleration_structure)`
1. `vkBeginCommandBuffer(cmdbuf)`
1. `vkCmdBindDescriptorSets(cmdbuf, ..., acceleration_set)`
2. `vkEndCommandBuffer(cmdbuf)`
3. `vkDestroyAccelerationStructureKHR(acceleration_structure)`
4. `vkCreateAccelerationStructureKHR(&acceleration_structure)`
5. `vkUpdateDescriptorSets(acceleration_set[0] = acceleration_structure)`
6. `vkQueueSubmit2(..., cmdbuf)` - validation error happens here

The full error looks like this:

> ERROR & VALIDATION => Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidCommandBuffer-VkAccelerationStructureKHR ] Object 0: handle = 0xea035600000000a7, type = VK_OBJECT_TYPE_ACCELERATION_STRUCTURE_KHR; Object 1: handle = 0x603cd90000000044, name = acceleration_set, type = VK_OBJECT_TYPE_DESCRIPTOR_SET; Object 2: handle = 0x20000258c50, name = CommandUtil CB, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xc659907c | You are adding vkQueueSubmit2 to VkCommandBuffer 0x20000258c50[CommandUtil CB] that is invalid because bound VkAccelerationStructureKHR 0xea035600000000a7[] was destroyed.
>    Object[0] - Type ACCELERATION_STRUCTURE_KHR, Value 0xea035600000000a7, Name "Unknown name"
>    Object[1] - Type DESCRIPTOR_SET, Value 0x603cd90000000044, Name "acceleration_set"
>    Object[2] - Type COMMAND_BUFFER, Value 0x20000258c50, Name "CommandUtil CB"

**Valid Usage ID**
`UNASSIGNED-CoreValidation-DrawState-InvalidCommandBuffer-VkAccelerationStructureKHR `

**Environment:**
 - OS: Windows 11 amd64
 - GPU: GeForce 3080, driver: 511.72
 - SDK or header version if building from repo: 1.3.204.0
 - Options enabled (synchronization, best practices, etc.): Just `VK_LAYER_KHRONOS_validation`

### LINKED ISSUE #3977: Potentially incorrect use-after-free VUIDs with bindless
We've recently started to need to silence a lot of validation layer warnings about objects potentially being in use after destroying them.

A subset of the errors we're ignoring;
```
         "UNASSIGNED-CoreValidation-DrawState-InvalidCommandBuffer-VkImageView",
        "VUID-vkDestroyAccelerationStructureKHR-accelerationStructure-02442",
        "VUID-vkDestroyBuffer-buffer-00922",
        "VUID-vkDestroyDevice-device-00378",
        "VUID-vkDestroyImageView-imageView-01026", 
```

Since it looks like there are no actual tests for this with a bindless setup, I'd like to ask you to investigate this as well as us doing our own investigation.

### PR DESCRIPTION
**Pull request by ziga-lunarg**
Closes #3835 
Closes #3977 
Closes #3729

### CODE REVIEW

File: layers/descriptor_sets.cpp [hunk 1]
```diff
@@ -2748,37 +2748,41 @@ bool CoreChecks::ValidateImageUpdate(VkImageView image_view, VkImageLayout image
...
-static void ReplaceStatePtr(DescriptorSet *set_state, T &dst, const T &src) {
+static void ReplaceStatePtr(DescriptorSet *set_state, T &dst, const T &src, bool is_update_after_bind) {
     if (dst) {
         dst->RemoveParent(set_state);
     }
     dst = src;
-    if (dst) {
+    // For descriptor bindings with UPDATE_AFTER_BIND only set the object as a child, but not the descriptor as a parent, so
+    // that destroying the object wont invalidate the descriptor
+    if (dst && !is_update_after_bind) {
```
Comment (ziga-lunarg) [inline layers/descriptor_sets.cpp:2758]: This is the only real change, for descriptors that use UPDATE_AFTER_BIND, don't set the descriptor set as parent, so when the resource gets destroyed the parent wont be invalidated.

**Review by jeremyg-lunarg — COMMENTED** [review]
I'm going to run the perf tests on this, just in case.

File: layers/descriptor_sets.cpp (hunk 1 above)
Comment (Tony-LunarG) [reply layers/descriptor_sets.cpp:2758]: Would the same apply to partially bound and #3729?

Comment (ziga-lunarg) [reply layers/descriptor_sets.cpp:2758]: I will write a test and try

Comment (ziga-lunarg) [reply layers/descriptor_sets.cpp]: Seems to work

**Comment by jeremyg-lunarg** [comment]
I got weird perf results against the old version of this PR yesterday. I'm not sure I trust current state of the CI machine. Going to rerun the new version today.

File: layers/descriptor_sets.cpp [hunk 2]
```diff
@@ -2748,37 +2748,40 @@ bool CoreChecks::ValidateImageUpdate(VkImageView image_view, VkImageLayout image
...
-static void ReplaceStatePtr(DescriptorSet *set_state, T &dst, const T &src) {
+static void ReplaceStatePtr(DescriptorSet *set_state, T &dst, const T &src, uint32_t binding) {
     if (dst) {
         dst->RemoveParent(set_state);
     }
     dst = src;
-    if (dst) {
+    // For descriptor bindings with UPDATE_AFTER_BIND or PARTIALLY_BOUND only set the object as a child, but not the descriptor as a
+    // parent, so that destroying the object wont invalidate the descriptor
+    if (dst && !set_state->IsUpdateAfterBind(binding) && !set_state->IsPartiallyBound(binding)) {
```
Comment (jeremyg-lunarg) [inline layers/descriptor_sets.cpp:2758]: each of these results in an unordered_map lookup to find the flags. I'm not sure if the compiler is smart enough to optimize them into one lookup.

Comment (jeremyg-lunarg) [inline layers/descriptor_sets.cpp:2758]: each of these results in an unordered_map lookup to find the flags. I'm not sure if the compiler is smart enough to optimize them into one lookup.

File: layers/descriptor_sets.cpp [hunk 3]
```diff
@@ -2252,7 +2252,7 @@ void cvdescriptorset::DescriptorSet::PerformCopyUpdate(ValidationStateTracker *d
         auto *src = src_set->descriptors_[src_start_idx + di].get();
         auto *dst = descriptors_[dst_start_idx + di].get();
         if (src->updated) {
-            dst->CopyUpdate(this, state_data_, src);
+            dst->CopyUpdate(this, state_data_, src, update);
```
Comment (jeremyg-lunarg) [inline layers/descriptor_sets.cpp:2255]: see line 2264 below. You could grab the flags for the update at this level and either pass them in here, or pass in an 'is_bindless' boolean.

Comment (jeremyg-lunarg) [inline layers/descriptor_sets.cpp:2255]: see line 2264 below. You could grab the flags for the update at this level and either pass them in here, or pass in an 'is_bindless' boolean.

Comment (jeremyg-lunarg) [reply layers/descriptor_sets.cpp:2255]: there's similar flag checking in PeformWriteUpdate() too.

Comment (jeremyg-lunarg) [reply layers/descriptor_sets.cpp:2255]: there's similar flag checking in PeformWriteUpdate() too.

**Review by jeremyg-lunarg — COMMENTED** [review]
This PR seems to have very minor performance regressions in several test cases. Not a huge problem but it should cause the code to be doing less work in many of those tests.

**Review by MarijnS95 — APPROVED** [review]
Thanks! This indeed solves our issue described in #3977.

File: layers/descriptor_sets.cpp [hunk 4]
```diff
@@ -2748,37 +2748,40 @@ bool CoreChecks::ValidateImageUpdate(VkImageView image_view, VkImageLayout image
 // correctly managing links to the parent DescriptorSet.
 // src and dst are shared pointers.
 template <typename T>
-static void ReplaceStatePtr(DescriptorSet *set_state, T &dst, const T &src) {
+static void ReplaceStatePtr(DescriptorSet *set_state, T &dst, const T &src, uint32_t binding) {
     if (dst) {
```
[the remark this answers is not in the document]
Comment (ziga-lunarg) [reply layers/descriptor_sets.cpp:2752]: By bindless do you mean `VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT`?

[the remark this answers is not in the document]
Comment (jeremyg-lunarg) [reply layers/descriptor_sets.cpp:2752]: or UPDATE_AFTER_BIND. The same conditions you have for AddParent(). If the descriptor never gets added, it seems like useless work to try to remove it.

**Comment by jeremyg-lunarg** [comment]
re-running perf on the newest version of this PR

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM. Perf hit is gone in most cases with the latest version.

**Pull request merged**